pubsublite/internal/wire/service_test.go

```diff
@@ -0,0 +1,519 @@
+				t.Errorf("service.unsafeCheckServiceStatus(): got (%v), want (%v)", gotErr, tc.wantErr)
+			}
+		})
+	}
+}
+
+func TestServiceAddRemoveStatusChangeReceiver(t *testing.T) {
+	receiver1 := newTestStatusChangeReceiver("receiver1")
+	receiver2 := newTestStatusChangeReceiver("receiver2")
+	receiver3 := newTestStatusChangeReceiver("receiver3")
+
+	service := new(testService)
+	service.AddStatusChangeReceiver(receiver1.Handle(), receiver1.OnStatusChange)
+	service.AddStatusChangeReceiver(receiver2.Handle(), receiver2.OnStatusChange)
+	service.AddStatusChangeReceiver(receiver3.Handle(), receiver3.OnStatusChange)
+
+	t.Run("All receivers", func(t *testing.T) {
+		service.UpdateStatus(serviceActive, nil)
+
+		receiver1.VerifyStatus(t, serviceActive)
+		receiver2.VerifyStatus(t, serviceActive)
+		receiver3.VerifyStatus(t, serviceActive)
+	})
+
+	t.Run("receiver1 removed", func(t *testing.T) {
+		service.RemoveStatusChangeReceiver(receiver1.Handle())
+		service.UpdateStatus(serviceTerminating, nil)
+
+		receiver1.VerifyNoStatusChanges(t)
+		receiver2.VerifyStatus(t, serviceTerminating)
+		receiver3.VerifyStatus(t, serviceTerminating)
+	})
+
+	t.Run("receiver2 removed", func(t *testing.T) {
+		service.RemoveStatusChangeReceiver(receiver2.Handle())
+		service.UpdateStatus(serviceTerminated, nil)
+
+		receiver1.VerifyNoStatusChanges(t)
+		receiver2.VerifyNoStatusChanges(t)
+		receiver3.VerifyStatus(t, serviceTerminated)
+	})
+}
+
+type testCompositeService struct {
+	receiver *testStatusChangeReceiver
+	compositeService
+}
+
+func newTestCompositeService(name string) *testCompositeService {
+	receiver := newTestStatusChangeReceiver(name)
+	ts := &testCompositeService{receiver: receiver}
+	ts.AddStatusChangeReceiver(receiver.Handle(), receiver.OnStatusChange)
+	ts.init()
+	return ts
+}
+
+func (ts *testCompositeService) AddServices(services ...service) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	ts.unsafeAddServices(services...)
+}
+
+func (ts *testCompositeService) RemoveService(service service) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	ts.unsafeRemoveService(service)
+}
+
+func (ts *testCompositeService) DependenciesLen() int {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	return len(ts.dependencies)
+}
+
+func (ts *testCompositeService) RemovedLen() int {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	return len(ts.removed)
+}
+
+func TestCompositeServiceNormalStop(t *testing.T) {
+	child1 := newTestService("child1")
+	child2 := newTestService("child2")
+	child3 := newTestService("child3")
+	parent := newTestCompositeService("parent")
+	parent.AddServices(child1, child2)
+
+	t.Run("Starting", func(t *testing.T) {
+		parent.Start()
+
+		child1.receiver.VerifyStatus(t, serviceStarting)
+		child2.receiver.VerifyStatus(t, serviceStarting)
+		parent.receiver.VerifyStatus(t, serviceStarting)
+
+		// child3 is added after Start() and should be automatically started.
+		parent.AddServices(child3)
+		child3.receiver.VerifyStatus(t, serviceStarting)
```

Comment:
Perhaps check that the status of child3 is `serviceUnitialized` prior to adding to the parent since I noticed that case hasn't been covered in other tests

Reply:
Good point. Done.